modules/storages/lib/api/v3/storage_files/storage_files_api.rb

```diff
@@ -26,51 +26,42 @@
 # See COPYRIGHT and LICENSE files for more details.
 #++
 
-module API::V3::FileLinks
-  class ParseCreateParamsService < ::API::ParseResourceParamsService
-    MAX_ELEMENTS = 20
-
-    attr_reader :request_body
-
-    private
-
-    def parse_attributes(request_body)
-      @request_body = request_body
-      assert_valid_elements
-
-      elements.map do |element|
-        super(element)
+module API::V3::StorageFiles
+  class StorageFilesAPI < ::API::OpenProjectAPI
+    helpers do
+      def handle_files_error(files)
+        case files.result
+        when :not_found
+          raise API::Errors::NotFound.new
+        when :not_authorized
+          raise API::Errors::Unauthorized.new
+        else
+          raise API::Errors::InternalError.new
+        end
       end
     end
 
-    def elements
-      @elements ||= request_body.dig("_embedded", "elements")
-    end
-
-    def assert_valid_elements
-      assert_elements_is_present
-      assert_elements_is_an_array
-      assert_elements_does_not_exceed_maximum
-    end
-
-    def assert_elements_is_present
-      return if elements.present?
+    resources :files do
+      get do
+        ::API::V3::Utilities::StorageRequests
+          .new(storage: @storage)
+          .files_query(user: current_user)
+          .match(
+            on_success: ->(query) {
+              files = query
+                        .call
+                        .on_failure { |r| handle_files_error(r) }
+                        .result
 
-      raise API::Errors::PropertyMissingError.new('_embedded/elements')
-    end
-
-    def assert_elements_is_an_array
-      return if elements.is_a?(Array)
-
-      raise API::Errors::PropertyFormatError.new('_embedded/elements', 'Array', elements.class.name)
-    end
-
-    def assert_elements_does_not_exceed_maximum
-      return if elements.size <= MAX_ELEMENTS
-
-      raise API::Errors::Validation.new('_embedded/elements',
-                                        I18n.t('api_v3.errors.too_many_elements_created_at_once',
-                                               max: MAX_ELEMENTS, actual: elements.size))
+              ::API::V3::StorageFiles::StorageFileCollectionRepresenter.new(
+                files,
+                self_link: api_v3_paths.storage_files(@storage.id),
+                current_user:
+              )
+            },
+            on_failure: ->(error) { handle_files_error(error) }
```

Comment:
According to the implementation of `match`, `error` is the actual `ServiceResult`s `result` value.

`match` connects the `ServiceResult` back to the "real" world, so to say.

Reply:
This is not the level I was taking about. What I would propose doing is change the line to 

```
on_failure: ->(error) { handle_files_error(error.result) }
```

That way, `handle_files_error` can be simplified to just take the errors. So semantically it could be renamed to `handle_error`. 

Now one could argue that there is a coupling between the error codes, e.g. `:not_found`, handling in here and the error codes being produced in [nextcloud_storage_query.rb](https://github.com/opf/openproject/pull/11315/files/69ca95a0afd51386f29cc2f137f3b3eef6b4e45a#diff-5e932046d0752a39fb69ce1f61ee5a9b4ff59ac022d5c99da882f8edc923c367R50-R58) which is a fair point and that might be an argument for keeping the naming. On the other hand, I think it shows that the coupling as such should be avoided. Once more storages are added, they would have to join into the same coupling. 

In the rest of the API there is already a handling of errors in place based on the [`create_and_merge_errors`](https://github.com/opf/openproject/blob/dev/lib/api/errors/error_base.rb#L43-L63) method. Admittedly, that is a different form of coupling but as that is used throughout the application one could also call it a convention. Of course, that is based on using `ActiveModel::Error`. So this would require moving towards having the error messages in the `errors` of a `ServiceResult`. For the time being, the `ActiveModel::Error` could also be in the `result` which would allow passing the result to the default error handing method (if the right codes are used).

Not related to the PR, it would make sense to move those convention error codes to a constant, that can then be referenced throughout the application.

Reply:
I got you wrong at first. Sorry. You actually spotted a bug there. I chose to pass the whole ˋServiceResultˋ into the error handler method because I'm using the original ˋon_failureˋ method which takes a block that gets the ˋServiceResultˋ in line 55.

The one below in line 64 is the ˋon_failureˋ of the new ˋmatchˋ which works differently, as explained before. Somehow this case is not in the spec I wrote, because it would fail in the error handler because the unwrapped error symbol has no ˋresultˋ method...

I will fix this in a concise way for sure!

Reply:
Nope, wrong again. That way of result handling is not working out. You are right. It could be simplified with your suggestion. I would still want to try something easier.